<center>
<img src="images/ods_stickers.jpg" />
    
## Introducción al Machine Learning

Basado en material de  [Yury Kashnitsky](https://yorko.github.io). Traducido y editado al español por [Ana Georgina Flesia](https://www.linkedin.com/in/georginaflesia/). Este material esta sujeto a los términos y condiciones de la licencia  [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Se permite el uso irrestricto para todo propósito no comercial.

# <center> Ejercicio 1. Análisis exploratorio con Pandas

<img src="images/pandas.jpg"  width=50% />


**En esta tarea se tiene que usar Pandas para responder preguntas sobre el  dataset [Adult](https://archive.ics.uci.edu/ml/datasets/Adult). **

Variables y su tipo:
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [2]:
import numpy as np
import pandas as pd
pd.set_option('display.max.columns', 100)
# para dibujar gráficos en jupyter notebook
%matplotlib inline 
import matplotlib.pyplot as plt
import seaborn as sns
# no nos gunstan los avisos
# se puden comentar las dos lineas si se quiere tener avisos
import warnings
warnings.filterwarnings('ignore')

In [3]:
data = pd.read_csv('adult.data.csv')
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [4]:
print(data.info())
print(data.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  salary          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
None
(32561, 15)


**1. Cuantas mujeres y hombres (variable *sex* ) estan representadas en este dataset?** 

In [5]:
#data.groupby(['sex']).
#male = data[data["sex"] == 'Male'] #El primer data es para saber a que data frame me refiero y el segundo me sirve para localizar la columna
#female = data[data["sex"] == 'Female']
data['sex'].value_counts()

Male      21790
Female    10771
Name: sex, dtype: int64

De la primera manera tengo que definir las cosas, no estan mal, mucho menos para aprender. Sin embargo, de la segunda manera puedo usar la función directa sin necesidad de definir nada

**2. Cual es la edad promedio (variable *age* ) de las mujeres?**

In [6]:
#female['age'].mean() Acá use el female definido
data[data["sex"] == 'Female']['age'].mean() #Acá no necesito definir nada

36.85823043357163

Al igual que en el inciso anterior, puedo definir, pero mejor si no lo hago asi no arrastro definiciones.

**3. Cual es el porcentaje de ciudadanos alemanes (variable *native-country*)?**

In [7]:
#data[data['native-country'] == 'Germany']['native-country'].value_counts(normalize=True)
data['native-country'].value_counts(normalize=True)['Germany']

0.004207487485028101

In [8]:
data['native-country'].value_counts(normalize=True).shape

(42,)

Notar que cuando aplico la función value_counts me da un data frame con la forma (n,) donde n es la cantidad de elementos diferentes en la columna sobre la cuál estoy trabajando.

**4. Cuales son la media y la desviacion estandard de la edad para aquellos que ganan mas de 50k por año (variable *salary*) ?**

In [9]:
data['salary'].value_counts()

<=50K    24720
>50K      7841
Name: salary, dtype: int64

In [10]:
data[data["salary"] == '>50K']['age'].mean() #COPIAR BIEN LAS CATEGORÍAS INCLUYENDO MAYÚSCULAS!!

44.24984058155847

In [11]:
data[data["salary"] == '>50K']['age'].std()

10.51902771985177

**5. Cuales son la media y la desviacion estandard de la edad para aquellos que ganan menos de 50k por año (variable *salary*)?**

In [12]:
data[data["salary"] == '<=50K']['age'].mean()

36.78373786407767

In [13]:
data[data["salary"] == '<=50K']['age'].std()

14.020088490824813

**6. Es cierto que la gente que gana mas que 50K tiene al menos educacion media completa? (variable *education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate*)**

In [14]:
array = ['Bachelors', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', 'Masters', 'Doctorate']
data.loc[(data['salary'] == '>50K') & data['education'].isin(array)]

age         workclass  fnlwgt   education  education-num  \
8       31           Private   45781     Masters             14   
9       42           Private  159449   Bachelors             13   
11      30         State-gov  141297   Bachelors             13   
14      40           Private  121772   Assoc-voc             11   
19      43  Self-emp-not-inc  292175     Masters             14   
...    ...               ...     ...         ...            ...   
32536   34           Private  160216   Bachelors             13   
32538   38           Private  139180   Bachelors             13   
32539   71                 ?  287372   Doctorate             16   
32545   39         Local-gov  111499  Assoc-acdm             12   
32554   53           Private  321865     Masters             14   

           marital-status       occupation   relationship                race  \
8           Never-married   Prof-specialty  Not-in-family               White   
9      Married-civ-spouse  Exec-managerial        Husband               White   
11     Married-civ-spouse   Prof-specialty        Husband  Asian-Pac-Islander   
14     Married-civ-spouse     Craft-repair        Husband  Asian-Pac-Islander   
19               Divorced  Exec-managerial      Unmarried               White   
...                   ...              ...            ...                 ...   
32536       Never-married  Exec-managerial  Not-in-family               White   
32538            Divorced   Prof-specialty      Unmarried               Black   
32539  Married-civ-spouse                ?        Husband               White   
32545  Married-civ-spouse     Adm-clerical           Wife               White   
32554  Married-civ-spouse  Exec-managerial        Husband               White   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
8      Female         14084             0              50  United-States   
9        Male          5178             0              40  United-States   
11       Male             0             0              40          India   
14       Male             0             0              40              ?   
19     Female             0             0              45  United-States   
...       ...           ...           ...             ...            ...   
32536  Female             0             0              55  United-States   
32538  Female         15020             0              45  United-States   
32539    Male             0             0              10  United-States   
32545  Female             0             0              20  United-States   
32554    Male             0             0              40  United-States   

      salary  
8       >50K  
9       >50K  
11      >50K  
14      >50K  
19      >50K  
...      ...  
32536   >50K  
32538   >50K  
32539   >50K  
32545   >50K  
32554   >50K  

[4535 rows x 15 columns]

Esto está mal, porque precisamente estoy viendo las personas que ganas mas de 50K al año y tienen al menos educación media completa. Podria contarlos y compararlos con los que ganan mas de 50K al año sin filtro alguno pero esto suena muy ineficiente. 
Si lo puedo usar para comprobar el resultado, que como se puede ver, me debería dar que hay $(7841-4535)=3306$ personas que ganan mas de 50K al año sin tener al menos el nivel medio de educación completo.

In [15]:
array = ['Bachelors', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', 'Masters', 'Doctorate']
data.loc[(data['salary'] == '>50K') & ~data['education'].isin(array)]

age         workclass  fnlwgt     education  education-num  \
7       52  Self-emp-not-inc  209642       HS-grad              9   
10      37           Private  280464  Some-college             10   
27      54                 ?  180211  Some-college             10   
38      31           Private   84154  Some-college             10   
55      43           Private  237993  Some-college             10   
...    ...               ...     ...           ...            ...   
32510   39           Private  107302       HS-grad              9   
32518   57         Local-gov  110417       HS-grad              9   
32519   46           Private  364548  Some-college             10   
32557   40           Private  154374       HS-grad              9   
32560   52      Self-emp-inc  287927       HS-grad              9   

           marital-status         occupation relationship                race  \
7      Married-civ-spouse    Exec-managerial      Husband               White   
10     Married-civ-spouse    Exec-managerial      Husband               Black   
27     Married-civ-spouse                  ?      Husband  Asian-Pac-Islander   
38     Married-civ-spouse              Sales      Husband               White   
55     Married-civ-spouse       Tech-support      Husband               White   
...                   ...                ...          ...                 ...   
32510  Married-civ-spouse     Prof-specialty      Husband               White   
32518  Married-civ-spouse       Craft-repair      Husband               White   
32519  Married-civ-spouse    Exec-managerial      Husband               White   
32557  Married-civ-spouse  Machine-op-inspct      Husband               White   
32560  Married-civ-spouse    Exec-managerial         Wife               White   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
7        Male             0             0              45  United-States   
10       Male             0             0              80  United-States   
27       Male             0             0              60          South   
38       Male             0             0              38              ?   
55       Male             0             0              40  United-States   
...       ...           ...           ...             ...            ...   
32510    Male             0             0              45              ?   
32518    Male         99999             0              40  United-States   
32519    Male             0             0              48  United-States   
32557    Male             0             0              40  United-States   
32560  Female         15024             0              40  United-States   

      salary  
7       >50K  
10      >50K  
27      >50K  
38      >50K  
55      >50K  
...      ...  
32510   >50K  
32518   >50K  
32519   >50K  
32557   >50K  
32560   >50K  

[3306 rows x 15 columns]

Entonces puedo ver que hay $3306$ personas que ganan más de 50K al año sin tener el nivel medio de ocupación completo

**7. Muestre los estadisticos para cada raza (*race* feature) y cada género (*sex* feature). Use *groupby()* y *describe()*. Encuentre la edad maxima de los hombres de la raza *Amer-Indian-Eskimo* .**

In [16]:
data.groupby(['race']).describe()

age                                                \
                      count       mean        std   min   25%   50%   75%   
race                                                                        
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  28.0  35.0  45.5   
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  28.0  36.0  45.0   
Black                3124.0  37.767926  12.759290  17.0  28.0  36.0  46.0   
Other                 271.0  33.457565  11.538865  17.0  25.0  31.0  41.0   
White               27816.0  38.769881  13.782306  17.0  28.0  37.0  48.0   

                           fnlwgt                                         \
                     max    count           mean            std      min   
race                                                                       
Amer-Indian-Eskimo  82.0    311.0  120831.147910   88338.636336  12285.0   
Asian-Pac-Islander  90.0   1039.0  159940.609240   85122.307505  14878.0   
Black               90.0   3124.0  228013.124200  123600.677422  19752.0   
Other               77.0    271.0  197124.191882   88856.775370  24562.0   
White               90.0  27816.0  187298.064280  103124.944196  18827.0   

                                                            education-num  \
                         25%       50%       75%        max         count   
race                                                                        
Amer-Indian-Eskimo   35036.5  102541.0  176142.0   445168.0         311.0   
Asian-Pac-Islander   93321.0  143098.0  194456.0   506329.0        1039.0   
Black               148212.0  205934.0  281693.0  1268339.0        3124.0   
Other               138103.0  188793.0  240091.0   481175.0         271.0   
White               116902.5  177627.0  233542.5  1484705.0       27816.0   

                                                                     \
                         mean       std  min  25%   50%   75%   max   
race                                                                  
Amer-Indian-Eskimo   9.311897  2.310387  2.0  9.0   9.0  10.0  16.0   
Asian-Pac-Islander  10.960539  2.811582  1.0  9.0  10.0  13.0  16.0   
Black                9.486236  2.297893  1.0  9.0   9.0  10.0  16.0   
Other                8.841328  3.226153  1.0  7.5   9.0  10.0  16.0   
White               10.135246  2.570307  1.0  9.0  10.0  13.0  16.0   

                   capital-gain                                                \
                          count         mean          std  min  25%  50%  75%   
race                                                                            
Amer-Indian-Eskimo        311.0   625.266881  2753.238961  0.0  0.0  0.0  0.0   
Asian-Pac-Islander       1039.0  1478.358037  9986.156906  0.0  0.0  0.0  0.0   
Black                    3124.0   609.940461  5139.653447  0.0  0.0  0.0  0.0   
Other                     271.0   934.660517  8625.128995  0.0  0.0  0.0  0.0   
White                   27816.0  1121.660375  7504.533302  0.0  0.0  0.0  0.0   

                            capital-loss                                   \
                        max        count       mean         std  min  25%   
race                                                                        
Amer-Indian-Eskimo  27828.0        311.0  34.176849  245.583106  0.0  0.0   
Asian-Pac-Islander  99999.0       1039.0  97.222329  423.556931  0.0  0.0   
Black               99999.0       3124.0  60.385083  337.394121  0.0  0.0   
Other               99999.0        271.0  61.070111  322.452705  0.0  0.0   
White               99999.0      27816.0  90.806155  410.833347  0.0  0.0   

                                     hours-per-week                        \
                    50%  75%     max          count       mean        std   
race                                                                        
Amer-Indian-Eskimo  0.0  0.0  1980.0          311.0  40.048232  11.695364   
Asian-Pac-Islander  0.0  0.0  2457.0  

Acá describo las edades dividiendolo por raza. Obviamente no me sirve mirar el máximo desde acá porque puede ser muy grande el caudal de mi categorización.

In [17]:
data.groupby(['race']).describe()['age']

count       mean        std   min   25%   50%   75%  \
race                                                                        
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  28.0  35.0  45.5   
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  28.0  36.0  45.0   
Black                3124.0  37.767926  12.759290  17.0  28.0  36.0  46.0   
Other                 271.0  33.457565  11.538865  17.0  25.0  31.0  41.0   
White               27816.0  38.769881  13.782306  17.0  28.0  37.0  48.0   

                     max  
race                      
Amer-Indian-Eskimo  82.0  
Asian-Pac-Islander  90.0  
Black               90.0  
Other               77.0  
White               90.0

In [18]:
print(data.groupby(['race']).describe()['age'].shape)

(5, 8)


In [19]:
data[data['race'] == 'Amer-Indian-Eskimo'].sort_values(by='age', ascending=False).head(n=1)['age']

12492    82
Name: age, dtype: int64

Esto no me convenció mucho, pero me dió el resultado directamente. Me dice que el cliente $N° 12492$ es las persona "Amer-Indian-Eskimo" con mayor edad (82 años).

La parte que no me convenció mucho es que le tuve que pedir muchas cosas por separado, primero que me seleccione los "Amer-Indian-Eskimo", luego que los ordene, luego que me muestre la primera fila y después que solo me muestre la edad. Debe haber algún comando directo que te pueda seleccionar el mayor valor de cierta columna, con cierto filtro.

In [20]:
i = data[data['race'] == 'Amer-Indian-Eskimo']['age'].idxmax()
print(i, data['age'][i])

12492 82


En este lo logré, tuve que definirlo e imprimirlo así porque si no no podía ver el número del cliente. Pero a fin de cuentas, de una manera más práctica pude saber cual era la persona con mayor edad de la raza "Amer-Indian-Eskimo"

In [21]:
data[data['race'].apply(lambda race: race == 'Amer-Indian-Eskimo')]['age'].max()

82

Otra manera aplicando el "apply"

In [22]:
data.groupby(['race'])['age'].agg([np.max])

amax
race                    
Amer-Indian-Eskimo    82
Asian-Pac-Islander    90
Black                 90
Other                 77
White                 90

In [23]:
data[data['race'] == 'Amer-Indian-Eskimo']['age'].agg([np.max])

amax    82
Name: age, dtype: int64

In [24]:
data.groupby(['sex']).describe()

age                                                       fnlwgt  \
          count       mean        std   min   25%   50%   75%   max    count   
sex                                                                            
Female  10771.0  36.858230  14.013697  17.0  25.0  35.0  46.0  90.0  10771.0   
Male    21790.0  39.433547  13.370630  17.0  29.0  38.0  48.0  90.0  21790.0   

                                                                               \
                 mean            std      min        25%       50%        75%   
sex                                                                             
Female  185746.311206  102986.078821  12285.0  117363.00  176077.0  228331.50   
Male    191771.449013  106740.702676  13769.0  118200.75  180140.0  241146.25   

                  education-num                                             \
              max         count       mean       std  min  25%   50%   75%   
sex                                                                          
Female  1484705.0       10771.0  10.035744  2.379954  1.0  9.0  10.0  12.0   
Male    1455435.0       21790.0  10.102891  2.662630  1.0  9.0  10.0  13.0   

             capital-gain                                                \
         max        count         mean          std  min  25%  50%  75%   
sex                                                                       
Female  16.0      10771.0   568.410547  4924.262944  0.0  0.0  0.0  0.0   
Male    16.0      21790.0  1329.370078  8326.312095  0.0  0.0  0.0  0.0   

                capital-loss                                              \
            max        count        mean         std  min  25%  50%  75%   
sex                                                                        
Female  99999.0      10771.0   61.187633  340.907518  0.0  0.0  0.0  0.0   
Male    99999.0      21790.0  100.213309  429.763097  0.0  0.0  0.0  0.0   

               hours-per-week                                               \
           max          count       mean        std  min   25%   50%   75%   
sex                                                                          
Female  4356.0        10771.0  36.410361  11.811300  1.0  30.0  40.0  40.0   
Male    3770.0        21790.0  42.428086  12.119755  1.0  40.0  40.0  49.0   

              
         max  
sex           
Female  99.0  
Male    99.0

**8. De las siguientes categorías, cual es la que tiene la mayor proporcion de ricos (>50K): married or single men (variable *marital-status*)? Considere como casados  aquellos que tienen *marital-status* comenzando por *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), el resto son bachelors.**

In [25]:
arr = ['Married-civ-spouse', 'Married-spouse-absent', 'Married-AF-spouse']
solt = data.loc[(data['salary'] == '>50K') & ~data['marital-status'].isin(arr)]
married = data.loc[(data['salary'] == '>50K') & data['marital-status'].isin(arr)]

In [26]:
print(solt['salary'].value_counts(), 'Solteros')
print(married['salary'].value_counts(), 'Casados')

>50K    1105
Name: salary, dtype: int64 Solteros
>50K    6736
Name: salary, dtype: int64 Casados


Me quedó feo como me tira el dato

**9. Cual es el máximo de horas que una persona trabaja por semana? (variable *hours-per-week*)? Cuantas personas trabajan ese numero de horas y cual es el porcentaje entre esas personas que ademas ganan mucho (>50K) ?**

In [27]:
# Entre su código python aquí

**10. Cuente el numero de horas de trabajo (*hours-per-week*) de aquellos que ganan poco, y de los que ganan mucho (*salary*) para cada pais (*native-country*). Cuales son esos conteos para Japon?**

In [28]:
data[data['native-country']== 'Japan'].groupby(['salary'])['hours-per-week'].describe(percentiles=[])

count       mean        std   min   50%   max
salary                                               
<=50K    38.0  41.000000  11.902759  10.0  40.0  65.0
>50K     24.0  47.958333  16.120414  21.0  42.5  99.0

In [2]:
data[data['age'] == data['age'].max()]

NameError: name 'data' is not defined